r2dbc/core/src/main/java/com/google/cloud/sql/core/CloudSqlConnection.java

```diff
@@ -0,0 +1,148 @@
+import java.time.Duration;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+/** {@link CloudSqlConnection} for connecting to Cloud SQL instances via R2DBC protocol. */
+public class CloudSqlConnection implements Connection {
+  private final Connection delegate;
+  private final ConnectionConfig config;
+
+  /** Creates an instance of Connection. */
+  public CloudSqlConnection(ConnectionConfig config, Connection delegate) {
+    this.config = config;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public Publisher<Void> setStatementTimeout(Duration timeout) {
+    return delegate.setStatementTimeout(timeout);
+  }
+
+  @Override
+  public Publisher<Void> beginTransaction() {
+    return delegate.beginTransaction();
+  }
+
+  @Override
+  public Publisher<Void> beginTransaction(TransactionDefinition definition) {
+    return delegate.beginTransaction(definition);
+  }
+
+  @Override
+  public Publisher<Void> close() {
+    return delegate.close();
+  }
+
+  @Override
+  public Publisher<Void> commitTransaction() {
+    return delegate.commitTransaction();
+  }
+
+  @Override
+  public Batch createBatch() {
+    return delegate.createBatch();
+  }
+
+  @Override
+  public Publisher<Void> createSavepoint(String name) {
+    return delegate.createSavepoint(name);
+  }
+
+  @Override
+  public Statement createStatement(String sql) {
+    return delegate.createStatement(sql);
+  }
+
+  @Override
+  public boolean isAutoCommit() {
+    return delegate.isAutoCommit();
+  }
+
+  @Override
+  public ConnectionMetadata getMetadata() {
+    return delegate.getMetadata();
+  }
+
+  @Override
+  public IsolationLevel getTransactionIsolationLevel() {
+    return delegate.getTransactionIsolationLevel();
+  }
+
+  @Override
+  public Publisher<Void> releaseSavepoint(String name) {
+    return delegate.releaseSavepoint(name);
+  }
+
+  @Override
+  public Publisher<Void> rollbackTransaction() {
+    return delegate.rollbackTransaction();
+  }
+
+  @Override
+  public Publisher<Void> rollbackTransactionToSavepoint(String name) {
+    return delegate.rollbackTransactionToSavepoint(name);
+  }
+
+  @Override
+  public Publisher<Void> setAutoCommit(boolean autoCommit) {
+    return delegate.setAutoCommit(autoCommit);
+  }
+
+  @Override
+  public Publisher<Void> setTransactionIsolationLevel(IsolationLevel isolationLevel) {
+    return delegate.setTransactionIsolationLevel(isolationLevel);
+  }
+
+  @Override
+  public Publisher<Boolean> validate(ValidationDepth depth) {
+    Publisher<Boolean> response = delegate.validate(depth);
+    Mono.from(response)
+        .doOnNext(
+            valid -> {
+              if (!valid) {
+                // Force refresh the connection info when it is invalid.
+                InternalConnectorRegistry.getInstance().forceRefresh(config);
+              }
+            })
+        .subscribeOn(Schedulers.boundedElastic())
```

Comment:
What is the significance of this line?

Reply:
To prevent blocking returning. I've added a comment. Please let me know if it is still not clear.

Reply:
OK thanks. Minor nit: could we move the comment to directly above this line just to make that point clear?

Reply:
Updated.